crates/sol-type-parser/src/lib.rs

```diff
@@ -83,3 +93,55 @@ pub(crate) fn str_parser<'a, E: ParserError<&'a str> + AddContext<&'a str, StrCo
         s.context(StrContext::Expected(StrContextValue::StringLiteral(s))),
     )
 }
+
+/// `( $( ${f()} ),* $(,)? )`
+pub(crate) fn tuple_parser<'a, O1, O2, E>(
+    f: impl Parser<&'a str, O1, E>,
+) -> impl Parser<&'a str, O2, E>
+where
+    O2: Accumulate<O1>,
+    E: ParserError<&'a str> + AddContext<&'a str, StrContext>,
+{
+    trace(
+        "tuple",
+        delimited(
+            (str_parser("("), space0),
+            cut_err(separated0(f, (str_parser(","), space0))),
+            (opt(","), space0, cut_err(str_parser(")"))),
+        ),
+    )
+}
+
+pub(crate) fn opt_ws_ident<'a>(input: &mut &'a str) -> PResult<Option<&'a str>> {
+    preceded(space0, opt(parse_identifier)).parse_next(input)
+}
+
+// Not public API.
+#[doc(hidden)]
+pub fn __internal_parse_signature<'a, const OUT: bool, F: Fn(ParameterSpecifier<'a>) -> T, T>(
```

Comment:
parsing code is always so cursed, think we could use docs in these internally for contributors, but no action required now